apps/api/src/domain/models/user-management/group/entities/invariant-validation/coscradUserGroupInvariantValidation.spec.ts

```diff
@@ -0,0 +1,162 @@
+    }
+
+    const invalidValuesAndDescriptions = Object.entries(fuzzData).reduce(
+        (acc: ValueAndDescription[], [key, value]: [FuzzDataType, unknown]) =>
+            validFuzzKeys.includes(key)
+                ? acc
+                : acc.concat({
+                      value,
+                      description: key,
+                  }),
+        []
+    );
+
+    if (isArray) {
+        const numberOfElementsInEachArray = 5;
+
+        return invalidValuesAndDescriptions.map(({ value, description }) => ({
+            description: `${description}[]`,
+            value: Array(numberOfElementsInEachArray).fill(value),
+        }));
+    }
+
+    return invalidValuesAndDescriptions;
+};
+
+const validGroupDto: DTO<CoscradUserGroup> = {
+    type: AggregateType.userGroup,
+    id: buildDummyUuid(),
+    label: 'test user group',
+    userIds: ['123'],
+    description: 'This is a test user group.',
+};
+
+// TODO [design] Consider making this a class.
+const createInvalidAggregateFactory =
+    <T extends DTO<Aggregate>>(validAggregate: T) =>
+    // TODO Support 'keys to exclude'
+    (overrides: { [K in keyof DTO<T>]?: unknown }): DTO<T> =>
+        ({
+            ...validAggregate,
+            ...overrides,
+        } as DTO<T>);
+
+const buildInvalidGroupDto = createInvalidAggregateFactory(validGroupDto);
+
+const dataSchema = getCoscradDataSchema(CoscradUserGroup);
+
+describe('CoscradUserGroup.validateInvariants', () => {
+    describe('when the data is valid', () => {
+        it('should return Valid', () => {
+            const validInstance = new CoscradUserGroup(validGroupDto);
+
+            const result = validInstance.validateInvariants();
+
+            expect(result).toBe(Valid);
+        });
+    });
+
+    describe('when the data is invalid', () => {
+        describe('when a simple invariant (generalized type) rule fails', () => {
+            Object.entries(dataSchema).forEach(([propertyName, propertySchema]) => {
+                describe(`when the property: ${propertyName} is invalid`, () => {
+                    const invalidValuesToUse = generateInvalidValuesForProperty(propertySchema);
+
+                    invalidValuesToUse.forEach(({ value, description }) => {
+                        describe(`when given the invalid value: ${JSON.stringify(
+                            value
+                        )} (${description})`, () => {
+                            it('should return the expected error', () => {
+                                const invalidDto = buildInvalidGroupDto({
+                                    id: value,
+                                });
+
+                                const invalidInstance = new CoscradUserGroup(invalidDto);
+
+                                const result = invalidInstance.validateInvariants();
+
+                                expect(result).toBeInstanceOf(InternalError);
```

Comment:
Can we sneak this into `assertCoscradDataTypeError`?

Reply:
It's not really necessary at all. I've updated the helper to take in `unknown` instead of `InternalError`.